mlflow/tracing/processor/otel_metrics_mixin.py

```diff
@@ -0,0 +1,123 @@
+
+_logger = logging.getLogger(__name__)
+
+
+class OtelMetricsMixin:
+    """
+    Mixin class that provides metrics recording capabilities for span processors.
+
+    This mixin is designed to be used with OpenTelemetry span processors to record
+    span-related metrics (e.g. duration) and metadata.
+    """
+
+    def __init__(self, *args, **kwargs):
+        """Initialize the mixin and pass through to parent classes."""
+        super().__init__(*args, **kwargs)
+        self._duration_histogram = None
+        self._trace_manager = InMemoryTraceManager.get_instance()
+
+    def _setup_metrics_if_necessary(self) -> None:
+        """
+        Set up OpenTelemetry metrics if not already configured previously.
+        """
+        if self._duration_histogram is not None:
+            return
+
+        endpoint = _get_otlp_metrics_endpoint()
+        if not endpoint:
+            return
+
+        protocol = _get_otlp_metrics_protocol()
+        if protocol == "grpc":
+            from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import (
+                OTLPMetricExporter,
+            )
+        elif protocol == "http/protobuf":
+            from opentelemetry.exporter.otlp.proto.http.metric_exporter import (
+                OTLPMetricExporter,
+            )
+        else:
+            _logger.warning(
+                f"Unsupported OTLP metrics protocol '{protocol}'. "
+                "Supported protocols are 'grpc' and 'http/protobuf'. "
+                "Metrics export will be skipped."
+            )
+            return
+
+        metric_exporter = OTLPMetricExporter(endpoint=endpoint)
+        reader = PeriodicExportingMetricReader(metric_exporter)
+        provider = MeterProvider(metric_readers=[reader])
+        metrics.set_meter_provider(provider)
+        meter = metrics.get_meter("mlflow.tracing")
+        self._duration_histogram = meter.create_histogram(
+            name="mlflow.trace.span.duration",
+            description="Duration of spans in milliseconds",
+            unit="ms",
+        )
+
+    def record_metrics_for_span(self, span: OTelReadableSpan) -> None:
+        """
+        Record metrics for a completed span.
+
+        This method should be called at the beginning of the on_end() method
+        to record span duration and associated metadata.
+
+        Args:
+            span: The completed OpenTelemetry span to record metrics for.
+        """
+        self._setup_metrics_if_necessary()
+
+        if self._duration_histogram is None:
+            return
+
+        span_type = span.attributes.get(SpanAttributeKey.SPAN_TYPE, SpanType.UNKNOWN)
+        try:
+            # Span attributes are JSON encoded by default; decode them for metric label readability
+            span_type = json.loads(span_type)
+        except (json.JSONDecodeError, TypeError):
+            pass
+
+        attributes = {
+            "root": str(span.parent is None),
```

Comment:
Ton confirm, this is correct?

Reply:
Yeah, if a span doesn't have a parent, it's the root

Reply:
`attributes` only accept strings?

Reply:
Ah, fair point - we don't technically have to stringify (boolean is supported). I'll remove str.